ols/src/cache/postgres_cache.py

```diff
@@ -0,0 +1,142 @@
+"""Cache that uses Postgres to store cached values."""
+
+import pickle
+from typing import Optional
+
+from langchain_core.messages.base import BaseMessage
+
+from ols.src.cache.cache import Cache
+
+
+class PostgresCache(Cache):
+    """Cache that uses Postgres to store cached values."""
+
+    CREATE_CACHE_TABLE = """
+        CREATE TABLE IF NOT EXISTS cache (
+            user_id         text NOT NULL,
+            conversation_id text NOT NULL,
+            key             text UNIQUE NOT NULL,
+            value           bytea,
+            updated_at      timestamp,
+            PRIMARY KEY(user_id, conversation_id)
+        );
+        """
+
+    CREATE_INDEX = """
+        CREATE INDEX IF NOT EXISTS timestamps
+            ON cache (updated_at)
+        """
+
+    SELECT_STATEMENT = """
+        SELECT value
+          FROM cache
+         WHERE user_id=%s and conversation_id=%s limit 1
+        """
+
+    UPDATE_STATEMENT = """
+        UPDATE cache
+           SET value=%s, updated_at=CURRENT_TIMESTAMP
+         WHERE user_id=%s AND conversation_id=%s
+        """
+
+    INSERT_STATEMENT = """
+        INSERT INTO cache(user_id, conversation_id, value, updated_at)
+        VALUES (%s, %s, %s, CURRENT_TIMESTAMP)
+        """
+
+    DELETE_STATEMENT = """
+        DELETE FROM cache
+         WHERE (user_id, conversation_id) in
+               (SELECT user_id, conversation_id FROM cache ORDER BY updated_at LIMIT
```

Comment:
The `LIMIT` value is missing?

Reply:
And maybe the closing `)` too?

Reply:
well not, it seems it can't be parametrized, so the string if formatted to add it

Reply:
I see